keanu-project/src/main/java/io/improbable/keanu/vertices/Vertex.java

```diff
@@ -99,7 +99,27 @@ public double logDensityAtValue() {
      * @return The value at this vertex after recalculating any parent non-probabilistic
      * vertices.
      */
-    public abstract T lazyEval();
+    public T lazyEval() {
+        Deque<Vertex<?>> stack = new ArrayDeque<>();
+        stack.push(this);
+        Set<Vertex<?>> hasCalculated = new HashSet<>();
+
+        while (!stack.isEmpty()) {
+
+            Vertex<?> head = stack.peek();
+            Set<Vertex<?>> parents = head.getParents();
+
+            if (parents.isEmpty() || areParentsCalculated(hasCalculated, parents) || head.isProbabilistic()) {
+                Vertex<?> top = stack.pop();
+                top.updateValue();
+                hasCalculated.add(top);
+            } else {
+                parents.forEach(stack::push);
```

Comment:
This can doubly add parents. Also, can we use a good ol for loop instead of the lambda? This line will be hit often enough that the lambda efficiency might be an issue.